### PR TITLE
Append trailing slash when the path is generated

### LINKED ISSUE #43219: Trailing slashes are not added when url has a dot
### Steps to reproduce
create a route that allows dots like in the documentation.

> By default, dynamic segments don't accept dots - this is because the dot is used as a separator for formatted routes. If you need to use a dot within a dynamic segment, add a constraint that overrides this – for example, id: /[^\/]+/ allows anything except a slash.

```ruby
get 'users/:username', to: 'users#show', constraints: { id: /[^\/]+/ }, as: :user, trailing_slash: true
```

call the path with username with and without a dot.

```ruby
user_path('someone')
user_path('some.one')
```
### Expected behavior
both should return  the url with trailing slash
```ruby
user_path('someone')
=> "/users/someone/"
user_path('some.one')
=> "/users/some.one/"
```

### Actual behavior
both should return  the url with trailing slash
```ruby
user_path('someone')
=> "/users/someone/"
user_path('some.one')
=> "/users/some.one"
```

Checking the code, theres a condition that explicitly asks if it has a dot it:

```ruby
def add_trailing_slash(path)
  if path.include?("?")
    path.sub!(/\?/, '/\&')
  elsif !path.include?(".") # here
    path.sub!(/[^\/]\z|\A\z/, '\&/')
  end
end
```

if this is not possible using trailing_slash , how can I achieve it? the problem is that Turbo is not being used if the url has a dot and it does not end in slash (see https://github.com/hotwired/turbo/issues/385)
### System configuration
**Rails version**: 6.1.4

**Ruby version**: 2.7.2-p137

### PR DESCRIPTION
**Pull request by casperisfine**
Fix: https://github.com/rails/rails/issues/43219

Instead of trying to parse the path later to insert the slash or bail out if the path look like it has a format indicator.

**This patch is based on the assumption that `url_for(path: '/blah')` is not an actual public API.** Which isn't something I'm certain of. But at the very least it doesn't seem documented.

### Git History

The `trailing_slash` option was added in https://github.com/rails/rails/commit/db4f421b0bfae226005591b1da90e4b42edf178d. However it was tacked on as a string operation on the return value of the URL Generator.

I don't understand why it wasn't added as an option of the generator itself.

The consequence is that then URL then have to be parsed to know where to insert the slash and wether it is safe or not to do.

https://github.com/rails/rails/pull/8704 fixed a few corner cases and added the following two tests:
```ruby
    assert_equal 'http://www.example.com/books/?q=code', url_for(trailing_slash: true, path: '/books?q=code')
    assert_equal 'http://www.example.com/books/?spareslashes=////', url_for(trailing_slash: true, path: '/books?spareslashes=////')
```

But as far as I can tell they don't correspond to any real behavior. Even if you use `url_for` with the implicit current request arguments as to normalize the current URL, the `path:` option is not passed and the URL generated again from `:controller` and `action`.

### CODE REVIEW

**Comment by silva96** [comment]
@casperisfine there's a small rubocop offense

File: actionpack/test/dispatch/request_test.rb [hunk 1]
```diff
@@ -42,9 +42,6 @@ class RequestUrlFor < BaseRequestTest
 
     assert_equal "/books", url_for(only_path: true, path: "/books")
 
-    assert_equal "http://www.example.com/books/?q=code", url_for(trailing_slash: true, path: "/books?q=code")
```
Comment (rafaelfranca) [inline actionpack/test/dispatch/request_test.rb:45]: Why were those tests removed?

Comment (casperisfine) [reply actionpack/test/dispatch/request_test.rb:45]: Because I believe they test something that's the underlying implementation an not public API https://github.com/rails/rails/pull/8704

Assuming that `url_for(path: "/foo")` isn't public API (it isn't documented as far as I can tell).

Comment (rafaelfranca) [reply actionpack/test/dispatch/request_test.rb:45]: But why not delete the test above that? I was not sure if it was because the behavior changed and the test failed, or just because we were cleaning up implementation detail tests.

In general we test non public API in the rails test suite if it is necessary, but if it is being tested using the public API it is fine to remove.

Comment (casperisfine) [reply actionpack/test/dispatch/request_test.rb:45]: > But why not delete the test above that as well? I was not sure if it was because the behavior changed and the test failed

Right. These two failed because they were testing the feature through a private API.

Comment (rafaelfranca) [reply actionpack/test/dispatch/request_test.rb:45]: So there is a change in behavior? I find that concerning. Would this change reintroduce the behavior #8704 was supposed to fix?

Comment (casperisfine) [reply actionpack/test/dispatch/request_test.rb:45]: No, I don't think so. But I feel like we're talking past each others, it's my fault I didn't take the time to properly braindump in the PR description, I'll do that.

Comment (casperisfine) [reply actionpack/test/dispatch/request_test.rb:45]: @rafaelfranca I added some more context to the PR and also added some more integration style tests.

**Review by rafaelfranca — APPROVED** [review]

File: actionpack/lib/action_dispatch/http/url.rb [hunk 2]
```diff
@@ -71,7 +71,8 @@ def path_for(options)
           path = options[:script_name].to_s.chomp("/")
           path << options[:path] if options.key?(:path)
 
-          add_trailing_slash(path) if options[:trailing_slash]
+          path = "/" if options[:trailing_slash] && path.blank?
```
Comment (ghiculescu) [inline actionpack/lib/action_dispatch/http/url.rb:74]: FYI https://github.com/rails/rails/issues/44373
